solve and unrad update

smichr opened the pull request:
closes #8780 as an alternative to that PR
fixes #8679 
fixes #8755 
fixes #8591 

MASTER:

``` python
>>> y=root(x,3);solve(y**2+y-x)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy\solvers\solvers.py", line 912, in solve
    solution = _solve(f[0], *symbols, **flags)
  File "sympy\solvers\solvers.py", line 1423, in _solve
    "\nNo algorithms are implemented to solve equation %s" % f)
NotImplementedError:
No algorithms are implemented to solve equation x**(2/3) + x**(1/3) - x
```

NOW:

```
>>> y=root(x,3);solve(y**2+y-x)
[0, (1/2 + sqrt(5)/2)**3]
```

smichr committed:
solve modifications

o move unrad to fallback

o keep only bases of denoms

  Although one might want the Pow form of a denominator when trying
  to find the domain of a function, it is not needed for zero testing.

o move hyperbolic rewriting out of _solve

o simplify which denominators are checked; checking solutions is time
  consuming, so if the number of expressions to check is reduced this
  will help keep solve from bogging down in that step. e.g. if x**3 and
  x appear as denominators, only the x needs to be checked.

o solve equations in the form 0**f(x) - 1

o always obtain a solution to cubics, etc... (issue 8755)

note:

The former solution to the expression in the test for issue 6644
came about by applying unrad to the eq to remove all radicals. This
produced a result that factored into a monomial, a quadratic and
an expression of degree 6. The monomial solution (0) is not valid
and only one of the quadratic solutions was valid. All 4 solutions
are now returned. For speed purposes, the spurious root is included
when checking is disabled.

smichr commented:
@nishithshah2211, included in the work here is a fix so that cubics always give a solution. Although what you have done is fine, the changes here would just erase your change. Would you be able to look over what has been done here?

nishithshah2211 commented on sympy/solvers/solvers.py:
```diff
@@ -694,6 +734,10 @@ def _sympified_list(w):
         elif isinstance(fi, (bool, C.BooleanAtom)) or fi.is_Relational:
             return reduce_inequalities(f, symbols=symbols)
```
Should we check the solution from `reduce_inequality` before returning it, like we are doing for other cases? Specifically, should we use `_simple_dens` or `denoms` to check if the denominators are zero for any value in the interval or solution and exclude that value from the interval? See [#8715] (https://github.com/sympy/sympy/issues/8715).

smichr replied:
Yes, checking should be done. Let's leave that to #8715, however. I just wanted to reorganize and enhance unrad in this PR. I've retracted most of the significant unrad changes to make review of this somewhat easier. Thanks for looking this over.

smichr committed:
unrad modifications

o issue 8679 has been addressed
o unrad no longer tries to handle powers with symbols of interest in
  the exponent
o unrad will now handle arbitrary radicals that appear in only two
  terms
o unrad (from solve) is only attempted for a symbol of interest; this
  leads to simpler expressions in the variable of interest.
o checking of solutions and simplifications are turned off while working
  within unrad
o denominators are no longer returned; if they are needed, they can be
  obtained in the calling environment
o the simple change-of-variables situation is better handled

smichr committed:
update wester V14 comment

smichr committed:
remove unused variables [skip ci]
